pymc_experimental/distributions/timeseries.py

```diff
@@ -0,0 +1,215 @@
+import warnings
+
+import numpy as np
+import pymc as pm
+import pytensor
+import pytensor.tensor as pt
+from pymc.distributions.dist_math import check_parameters
+from pymc.distributions.distribution import Distribution, SymbolicRandomVariable
+from pymc.distributions.logprob import ignore_logprob, logp
+from pymc.distributions.shape_utils import (
+    _change_dist_size,
+    change_dist_size,
+    get_support_shape_1d,
+)
+from pymc.logprob.abstract import _logprob
+from pymc.pytensorf import intX
+from pymc.util import check_dist_not_registered
+from pytensor.graph.basic import Node
+from pytensor.tensor import TensorVariable
+from pytensor.tensor.random.op import RandomVariable
+
+
+class DiscreteMarkovChainRV(SymbolicRandomVariable):
+    n_lags: int
+    default_output = 1
+    _print_name = ("DiscreteMC", "\\operatorname{DiscreteMC}")
+
+    def __init__(self, *args, n_lags, **kwargs):
+        self.n_lags = n_lags
+        super().__init__(*args, **kwargs)
+
+    def update(self, node: Node):
+        return {node.inputs[-1]: node.outputs[0]}
+
+
+class DiscreteMarkovChain(Distribution):
+    r"""
+    A Discrete Markov Chain is a sequence of random variables
+    .. math::
+        \{x_t\}_{t=0}^T
+    Where transition probability P(x_t | x_{t-1}) depends only on the state of the system at x_{t-1}.
+
+    Parameters
+    ----------
+    P: tensor
+        Matrix of transition probabilities between states. Rows must sum to 1.
+        One of P or P_logits must be provided.
+    P_logit: tensor, optional
+        Matrix of transition logits. Converted to probabilities via Softmax activation.
+        One of P or P_logits must be provided.
+    steps: tensor, optional
+        Length of the markov chain. Only needed if state is not provided.
+    init_dist : unnamed distribution, optional
```

Comment:
IIUC, this is a DiscreteMarkovChain in the sense that the emission distribution is a Categorical - if so you might want to restrict the init_dist to  pm.Categorical.dist? Or even better only allow kwarg of `p: Tensor = np.full(n_states, 1/n_states)`

Reply:
Categorical makes the most sense but in principle it could be any discrete distribution no?

Reply:
Depending on how you are going to deal with unbounded (Poisson) discrete distribution or component with different support - I dont quite see it from the docstring and your implementation, but that's something we can think about it more in follow up.

Reply:
Yes, as it stands it would allow you to use the unbounded init_dist, then fail on sampling when a state outside of the support came along. Is there a way to quickly check the support of a discrete distribution? In an earlier version I did a check like:

```python
x0_probs = init_dist.owner.inputs[-1].eval()
n_cats = 1 if x0_probs.ndim == 0 else len(x0_probs)
```

Which checked the number of categories in the categorical. The if/else logic was there to check for a Bernoulli, but it wouldn't catch a Poisson. 

Anyway, I think you suggestion is probably right, but my last doubt is that all the other distributions in `distributions.timeseries` use `init_dist` in their signature and expect an unnamed distribution, and I think there is value in consistency. I could add a check (and update the docstring) to enforce that the distribution passed must be categorical (for now at least).

Reply:
I tried to split the baby by adding a check that `init_dist` must be `pm.Categorical`, and it raises an error if not. I really wanted the API to be consistent with the other time series distributions. If you think that's not valuable, I'll change it so that `init_dist` becomes `init_p` instead.

Reply:
I am not sure about the categorical constraint. I might want to pass a diracdelta for posterior predictive or use a discrete uniform because it's simpler. Anyway, doesn't need to be a blocker.

Reply:
I'm glad you flagged it, I was about to push a change to let Bernoulli in, but the more I think about it the more I think it's better to just assume the user will be responsible?

Reply:
I removed the check.